main/lsp/ExtractVariable.cc

```diff
@@ -0,0 +1,239 @@
+            skipLoc(arg.loc());
+        }
+        updateEnclosingScope(tree, methodDef.rhs.loc());
+    }
+
+    void preTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &block = ast::cast_tree_nonnull<ast::Block>(tree);
+        for (auto &arg : block.args) {
+            skipLoc(arg.loc());
+        }
+        updateEnclosingScope(tree, block.body.loc());
+    }
+
+    void preTransformAssign(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &assign = ast::cast_tree_nonnull<ast::Assign>(tree);
+        skipLoc(assign.lhs.loc());
+    }
+
+    void preTransformIf(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &if_ = ast::cast_tree_nonnull<ast::If>(tree);
+        updateEnclosingScope(tree, if_.thenp.loc());
+        updateEnclosingScope(tree, if_.elsep.loc());
+    }
+
+    void preTransformRescue(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &rescue = ast::cast_tree_nonnull<ast::Rescue>(tree);
+        updateEnclosingScope(tree, rescue.body.loc());
+        updateEnclosingScope(tree, rescue.else_.loc());
+        updateEnclosingScope(tree, rescue.ensure.loc());
+    }
+
+    void preTransformRescueCase(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &rescueCase = ast::cast_tree_nonnull<ast::RescueCase>(tree);
+        updateEnclosingScope(tree, rescueCase.body.loc());
+        skipLoc(rescueCase.var.loc());
+        // TODO: skipLoc for exceptions
+    }
+
+    void preTransformWhile(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &while_ = ast::cast_tree_nonnull<ast::While>(tree);
+        updateEnclosingScope(tree, while_.body.loc());
+    }
+};
+
+vector<unique_ptr<TextDocumentEdit>> getExtractVariableEdits(LSPTypecheckerDelegate &typechecker,
+                                                             const LSPConfiguration &config,
+                                                             unique_ptr<Range> selectionRange,
+                                                             const core::Loc selectionLoc) {
+    auto loc = selectionLoc;
+    auto file = loc.file();
+    const core::GlobalState &gs = typechecker.state();
+    vector<unique_ptr<TextEdit>> edits;
+
+    ExtractVariableWalk extractVariableWalk(loc);
+    auto desugaredTree = typechecker.getDesugared(file);
+    core::Context ctx(gs, core::Symbols::root(), file);
+    ast::TreeWalk::apply(ctx, extractVariableWalk, desugaredTree);
+
+    if (extractVariableWalk.foundExactMatch()) {
+        vector<unique_ptr<TextEdit>> edits;
+
+        auto locOffsets = loc.offsets();
+        core::LocOffsets whereToInsert = core::LocOffsets::none();
+        auto enclosingScope = extractVariableWalk.enclosingScope;
+        // For all cases except InsSeq and ClassDef, extractVariableWalk.enclosingScopeLoc should be
+        // the same as what we're pulling out from the ExpressionPtr, but we'll just get it directly
+        // for consistency.
```

Comment:
Once we have the scope that the selection is inside, we need to walk the through the scope to find where to insert the new variable declaration. For one line scope, that's easy: we can put it at the start of the scope, since that will be before the selection. For `InsSeq` and `ClassDef`, we walk through each statement to find the statement that contains the selection, and insert just before that statement.